### PR TITLE
Introduce Arrays

### PR DESCRIPTION
**Pull request by abertschi**
We introduce Array behavior. Arrays in PHP are Map like. This PR does not fully
implement Arrays as they are specified.  But introduces an initial
implementation which works for most cases (and suffices our benchmark requirements)

Arrays are represented by different backends making use of Truffle Libraries for
polymorphic dispatch with specializations.

- [x] array()
- [X] array(val1, val2, ...)
- [X] support for nested arrays of long, or Object type (Object can store mixed values; long, boolean, double, or arrays itself)
- [X] arrays are initially backed by a long[] based implementation and generalized if needed
- [X] we support long[] and Object[] based backends
- [ ] Map like backends (for $A = array(); $A[-1] = ..., like behavior is not implemented yet)
  - [ ] No key/ value backend
- [X] array backends have initial capacity and are grown if capacity is exceeded.
- [ ] there is currently no support for exceptions if an unwritten field is read.
  - [ ] i.e. ... = $A[10000]; this may work if backend has capacity allocated for index.
  - [ ] PHP throws an error if an uninitialized index is accessed, this can easily be implemented for Object[] based backend.
  - [ ] for primitive backends we may have to keep track of initializations with bitmaps or different implementation.
  - [X] array is automatically grown if current capacity is reached.
    - If index is too large, transition to a Map like backend may be more performant  
- PHP has a copy by value behavior if an array is assigned to another value
  - [x] ~~we currently copy by reference~~ by value #30

### CODE REVIEW

File: graalphp-language/src/main/java/org/graalphp/nodes/array/ArrayWriteNode.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            },
+            limit = LIMIT)
+    protected PhpArray writeInBoundsSameType(
+            PhpArray array,
+            long index,
+            Object value,
+            @CachedLibrary("array.getBackend()") ArrayLibrary library) {
+        log("writeInBoundsSameType");
+
+        library.write(array.getBackend(), (int) index, value);
```
Comment (chumer) [inline graalphp-language/src/main/java/org/graalphp/nodes/array/ArrayWriteNode.java:60]: unsafe int cast.

File: graalphp-language/src/main/java/org/graalphp/nodes/array/ArrayWriteNode.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+ * Array write node uses variable array backends and generalizes if needed
+ *
+ * @author abertschi
+ */
+@NodeChild(value = "receiver")
+@NodeChild(value = "index")
+@NodeChild(value = "value")
+public abstract class ArrayWriteNode extends PhpExprNode {
+
+    private static final Logger L = PhpLogger.getLogger(ArrayWriteNode.class.getSimpleName());
```
Comment (chumer) [inline graalphp-language/src/main/java/org/graalphp/nodes/array/ArrayWriteNode.java:23]: Why did you not use TruffleLogger?

File: graalphp-language/src/main/java/org/graalphp/nodes/array/ArrayReadNode.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+@NodeChild(value = "index")
+public abstract class ArrayReadNode extends PhpExprNode {
+
+    private static final String LIMIT = ArrayLibrary.SPECIALIZATION_LIMIT;
+
+    // TODO: the smallest unit for integer numbers are long, java cant have long array length.
+    // we currently cast to int. for better error handling add error handling if not fit within int.
+    // return arrays.read(array.getBackend(), (int) index);
+
+    @Specialization(guards = "arrays.isArray(array.getBackend())", limit = LIMIT)
```
Comment (chumer) [inline graalphp-language/src/main/java/org/graalphp/nodes/array/ArrayReadNode.java:23]: isn't the check for PhpArray using the Java type implicitly enough to find out whether this is an array?

File: graalphp-language/src/main/java/org/graalphp/nodes/PhpNode.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package org.graalphp.nodes;
+
+import com.oracle.truffle.api.nodes.Node;
+
+/**
+ * @author abertschi
+ */
+public class PhpNode extends Node {
```
Comment (chumer) [inline graalphp-language/src/main/java/org/graalphp/nodes/PhpNode.java:8]: abstract?

File: graalphp-language/src/main/java/org/graalphp/nodes/array/ArrayWriteNode.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    /**
+     * write within existing capacity of different type => needs generalization
+     */
+    @Specialization(
+            guards = {
+                    "!library.acceptsValue(array.getBackend(), value)"
+                    , "isArrayInBounds(array, index)"
+            },
+            limit = LIMIT)
+    protected PhpArray writeInBoundsWrongType(
```
Comment (chumer) [inline graalphp-language/src/main/java/org/graalphp/nodes/array/ArrayWriteNode.java:73]: An important design choice here could be that library messages should be as close to language semantics and abstract the representation of the array (in php that is just read + write). You can then internally dispatch on the layout/representation. With this node you are not really abstracting the representation of the array, but just the storage which is only a little bit useful as its actually an implementation detail of the language. But it looks like a good start. 

You could do an ArrayLibrary + StorageLibrary and then let the caches merge by creating a @CachedLibrary to a final field.

File: graalphp-language/src/main/java/org/graalphp/types/PhpTypes.java [hunk 6]
```diff
@@ -3,14 +3,17 @@
...
 /**
  * @author abertschi
  */
 @TypeSystem({
         boolean.class,
         long.class,
-        double.class
+        double.class,
+        PhpArray.class,
+        Object[].class
```
Comment (chumer) [inline graalphp-language/src/main/java/org/graalphp/types/PhpTypes.java:16]: No need to register types here.

File: graalphp-language/src/main/java/org/graalphp/nodes/array/ArrayWriteNode.java (hunk 5 above)
Comment (abertschi) [reply graalphp-language/src/main/java/org/graalphp/nodes/array/ArrayWriteNode.java:73]: Thanks a lot for your pointers :+1: Will incorporate it :) Regarding this suggestion, I see the benefits of separation of concern of having a language semantic representation of the array and a technical. Are there other benefits apart from separation of concern, i.e. regarding performance?

**Pull request merged**